### PR TITLE
City region country entity

### PR DESCRIPTION
**Pull request by irasht**

### CODE REVIEW

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.softserve.actent.model.entity;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.*;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/City.java]: replace with single imports

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.softserve.actent.model.entity;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.*;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: replace with single imports

File: src/main/java/com/softserve/actent/model/entity/Region.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package com.softserve.actent.model.entity;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import org.hibernate.validator.constraints.Length;
+
+import javax.persistence.*;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: replace with single imports

File: src/main/java/com/softserve/actent/model/entity/Country.java (hunk 2 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: separate all imports. do not use *

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import javax.validation.constraints.NotBlank;
+
+@Data
+@NoArgsConstructor
+@Entity(name = "cities")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/City.java]: name of table must be in @Table annotation

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import javax.validation.constraints.NotBlank;
+import java.util.List;
+
+@Data
+@NoArgsConstructor
+@Entity(name = "countries")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: name of table must be in @Table annotation

File: src/main/java/com/softserve/actent/model/entity/Region.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import javax.validation.constraints.NotBlank;
+import java.util.List;
+
+@Data
+@NoArgsConstructor
+@Entity(name = "regions")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: name of table must be in @Table annotation

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+public class City {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "City can`t be blank")
+    @Length(max = 30, message = "Too long")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/City.java]: magic strings: create string constants in resources/StringConstants.java, replace with them

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+public class Country {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "Country can`t be blank")
+    @Length(max = 30, message = "Too long")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: magic strings: create string constants in resources/StringConstants.java, replace with them

File: src/main/java/com/softserve/actent/model/entity/Region.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+public class Region {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "Region can`t be blank")
+    @Length(max = 30, message = "Too long")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: magic strings: create string constants in resources/StringConstants.java, replace with them

File: src/main/java/com/softserve/actent/model/entity/City.java (hunk 1 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/City.java]: separate all imports. do not use *

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    @Column(unique = true, nullable = false, length = 30)
+    private String name;
+
+    @OneToMany(mappedBy = "country")
+    private List<Region> regions;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: replace with Set?

File: src/main/java/com/softserve/actent/model/entity/Region.java (hunk 3 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: separate all imports. do not use *

File: src/main/java/com/softserve/actent/model/entity/City.java (hunk 4 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/City.java]: add annotation  @Table with name cities. @Entity should not be with name

File: src/main/java/com/softserve/actent/model/entity/Country.java (hunk 5 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: add annotation  @Table with name countries. @Entity should not be with name

File: src/main/java/com/softserve/actent/model/entity/Region.java (hunk 6 above)
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: add annotation  @Table with name regionss. @Entity should not be with name

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+@Data
+@NoArgsConstructor
+@Entity(name = "cities")
+public class City {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "City can`t be blank")
```
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/City.java]: do not use magic String. please use StringConstants class

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+@Data
+@NoArgsConstructor
+@Entity(name = "countries")
+public class Country {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "Country can`t be blank")
```
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: do not use magic String. please use StringConstants class

File: src/main/java/com/softserve/actent/model/entity/Region.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+@Data
+@NoArgsConstructor
+@Entity(name = "regions")
+public class Region {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = "Region can`t be blank")
```
Comment (nazarmavko) [inline src/main/java/com/softserve/actent/model/entity/Region.java]: do not use magic String. please use StringConstants class

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
+package com.softserve.actent.model.entity;
+
+import com.softserve.actent.resources.StringConstants;
+import lombok.*;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/City.java]: replace with single imports

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
+package com.softserve.actent.model.entity;
+
+import com.softserve.actent.resources.StringConstants;
+import lombok.*;
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: replace with single imports

File: src/main/java/com/softserve/actent/model/entity/Country.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+@Entity
+@Table(name = "countries")
+public class Country {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = StringConstants.COUNTRY_SHOULD_NOT_BE_BLANK)
+    @Length(max = 30, message = "Too long")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Country.java]: magic string

**Review by md1guy — DISMISSED** [review]

**Review by nazarmavko — DISMISSED** [review]

**Review by Senchakmen — DISMISSED** [review]

File: src/main/java/com/softserve/actent/constant/StringConstants.java [hunk 17]
```diff
@@ -4,4 +4,9 @@
 
     public static final String EMPTY_REWIEW_TEXT = "Review text can not be empty.";
     public static final String NO_REVIEW_SCORE = "You must provide review score.";
+    public static final String CITY_SHOULD_NOT_BE_BLANK = "City can`t be blank";
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/constant/StringConstants.java:8]: provide empty string between constants for various classes

File: src/main/java/com/softserve/actent/constant/StringConstants.java [hunk 18]
```diff
@@ -4,4 +4,9 @@
...
+    public static final String REGION_SHOULD_NOT_BE_BLANK = "Region can`t be blank";
+    public static final String COUNTRY_SHOULD_NOT_BE_BLANK ="Country can`t be blank";
+    public static final String TOO_LONG ="Too long";
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/constant/StringConstants.java]: error message needs to be more specific

File: src/main/java/com/softserve/actent/model/entity/Location.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+import javax.validation.constraints.NotBlank;
+
+@Data
+@NoArgsConstructor
+@Entity
+@Table(name = "location")
```
Comment (md1guy) [inline src/main/java/com/softserve/actent/model/entity/Location.java]: names of tables should be in plural form

**Review by md1guy — DISMISSED** [review]

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NonNull
+    @NotBlank(message = StringConstants.CITY_SHOULD_NOT_BE_BLANK)
+    @Length(max = 30, message = StringConstants.CITY_NO_LONGER_THAT_THIRTY_SYMBOLS)
+    @Column(unique = true, nullable = false, length = 30)
+    private String name;
+
+
```
Comment (thebadpanda) [inline src/main/java/com/softserve/actent/model/entity/City.java]: remove one empty string

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    @ManyToOne
+    private Region region;
+}
```
Comment (thebadpanda) [inline src/main/java/com/softserve/actent/model/entity/City.java]: add newline at the end of file

File: src/main/java/com/softserve/actent/constant/StringConstants.java [hunk 22]
```diff
@@ -4,4 +4,14 @@
 
     public static final String EMPTY_REWIEW_TEXT = "Review text can not be empty.";
     public static final String NO_REVIEW_SCORE = "You must provide review score.";
+
+    public static final String CITY_SHOULD_NOT_BE_BLANK = "City can`t be blank";
+    public static final String REGION_SHOULD_NOT_BE_BLANK = "Region can`t be blank";
+    public static final String COUNTRY_SHOULD_NOT_BE_BLANK ="Country can`t be blank";
+    public static final String LOCATION_SHOULD_NOT_BE_BLANK ="Country can`t be blank";
+
+    public static final String LOCATION_NO_LONGER_THAT_FIFTY_SYMBOLS ="Location shouldn`t be longer than 50";
```
Comment (thebadpanda) [inline src/main/java/com/softserve/actent/constant/StringConstants.java]: THAT -> THAN

File: src/main/java/com/softserve/actent/model/entity/City.java [hunk 23]
```diff
@@ -26,11 +26,11 @@
...
-    @Length(max = 30, message = StringConstants.CITY_NO_LONGER_THAT_THIRTY_SYMBOLS)
+    @Length(max = 30, message = StringConstants.CITY_NO_LONGER_THAN_THIRTY_SYMBOLS)
     @Column(unique = true, nullable = false, length = 30)
     private String name;
 
-
     @ManyToOne
     private Region region;
+
 }
```
Comment (thebadpanda) [inline src/main/java/com/softserve/actent/model/entity/City.java]: add newline at the end of file. Under "}"

**Review by md1guy — APPROVED** [review]

**Review by Senchakmen — APPROVED** [review]

**Review by thebadpanda — APPROVED** [review]

**Pull request merged**